File: src/thoth/game/inventory.cpp
#include "thoth/game/inventory.hpp"

#include <algorithm>

namespace thoth::game {

bool Inventory::add(ItemId item, int count)
{
    if (item == ItemId::None || count <= 0) {
        return false;
    }

    for (auto& stack : stacks_) {
        if (stack.item == item) {
            stack.count += count;
            return true;
        }
    }

    stacks_.push_back(ItemStack{item, count});
    return true;
}

bool Inventory::consume(ItemId item, int count)
{
    if (!canConsume(item, count)) {
        return false;
    }

    for (auto& stack : stacks_) {
        if (stack.item == item) {
            stack.count -= count;
            break;
        }
    }

    stacks_.erase(
        std::remove_if(stacks_.begin(), stacks_.end(), [](const ItemStack& stack) {
            return stack.count <= 0;
        }),
        stacks_.end());
    return true;
}

bool Inventory::canConsume(ItemId item, int count) const
{
    if (item == ItemId::None || count <= 0) {
        return true;
    }
    return this->count(item) >= count;
}

bool Inventory::canConsumeAll(const std::vector<ItemStack>& stacks) const
{
    for (const auto& stack : stacks) {
        if (!canConsume(stack.item, stack.count)) {
            return false;
        }
    }
    return true;
}

bool Inventory::consumeAll(const std::vector<ItemStack>& stacks)
{
    if (!canConsumeAll(stacks)) {
        return false;
    }

    for (const auto& stack : stacks) {
        if (!consume(stack.item, stack.count)) {
            return false;
        }
    }
    return true;
}

int Inventory::count(ItemId item) const
{
    for (const auto& stack : stacks_) {
        if (stack.item == item) {
            return stack.count;
        }
    }
    return 0;
}

std::vector<ItemStack> Inventory::stacks() const
{
    std::vector<ItemStack> stable;
    for (const auto& def : itemDefs()) {
        const int amount = count(def.id);
        if (amount > 0) {
            stable.push_back(ItemStack{def.id, amount});
        }
    }
    return stable;
}

void Inventory::clear()
{
    stacks_.clear();
}

} // namespace thoth::game

File: src/game/inventory.lua
local Defs = require("src.game.defs")

local Inventory = {}
Inventory.__index = Inventory

function Inventory.new(stacks)
    local self = setmetatable({ counts = {} }, Inventory)
    if stacks then
        for _, stack in ipairs(stacks) do
            self:add(stack.item, stack.count)
        end
    end
    return self
end

function Inventory:add(item, count)
    if not item or count <= 0 then
        return false
    end
    if not Defs.item(item) then
        return false
    end
    self.counts[item] = (self.counts[item] or 0) + count
    return true
end

function Inventory:consume(item, count)
    if self:count(item) < count then
        return false
    end
    self.counts[item] = self.counts[item] - count
    if self.counts[item] <= 0 then
        self.counts[item] = nil
    end
    return true
end

function Inventory:canConsume(items)
    for item, count in pairs(items) do
        if self:count(item) < count then
            return false
        end
    end
    return true
end

function Inventory:consumeAll(items)
    if not self:canConsume(items) then
        return false
    end
    for item, count in pairs(items) do
        self:consume(item, count)
    end
    return true
end

function Inventory:count(item)
    return self.counts[item] or 0
end

function Inventory:firstMatching(items)
    for _, item in ipairs(items) do
        if self:count(item) > 0 then
            return item
        end
    end
    return nil
end

function Inventory:stacks()
    local stacks = {}
    for _, item in ipairs(Defs.itemOrder) do
        local count = self:count(item)
        if count > 0 then
            stacks[#stacks + 1] = { item = item, count = count }
        end
    end
    return stacks
end

function Inventory:clone()
    return Inventory.new(self:stacks())
end

return Inventory
